share/CMakeJS.cmake

```diff
@@ -0,0 +1,642 @@
+    if(VERBOSE)
+        message(STATUS "CMake Calling...")
+    endif()
+
+    # Check for cmake-js installations
+    find_program(CMAKE_JS_EXECUTABLE
+      NAMES "cmake-js" "cmake-js.exe"
+      PATHS "$ENV{PATH}" "$ENV{ProgramFiles}/cmake-js"
+      DOC "cmake-js system executable binary"
+      REQUIRED
+    )
+    if (NOT CMAKE_JS_EXECUTABLE)
+        message(FATAL_ERROR "cmake-js system installation not found! Please run 'npm -g install cmake-js@latest' and try again.")
+        return()
+    endif()
+
+    _cmakejs_normalize_path(CMAKE_JS_EXECUTABLE)
+    string(REGEX REPLACE "[\r\n\"]" "" CMAKE_JS_EXECUTABLE "${CMAKE_JS_EXECUTABLE}")
+
+    find_program(CMAKE_JS_NPM_PACKAGE
+      NAMES "cmake-js" "cmake-js.exe"
+      PATHS "${CMAKE_CURRENT_SOURCE_DIR}/node_modules/cmake-js/bin"
+      DOC "cmake-js project-local npm package binary"
+      REQUIRED
+    )
+    if (NOT CMAKE_JS_NPM_PACKAGE)
+        message(FATAL_ERROR "cmake-js project-local npm package not found! Please run 'npm install' and try again.")
+        return()
+    endif()
+
+    # Execute the CLI commands, and write their outputs into the cached vars
+    # where the remaining build processes expect them to be...
+    execute_process(
+      COMMAND "${CMAKE_JS_EXECUTABLE}" "print-cmakejs-include" "--log-level error" "--generator ${CMAKE_GENERATOR}"
+      WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+      OUTPUT_VARIABLE CMAKE_JS_INC
+    )
+
+    execute_process(
+      COMMAND "${CMAKE_JS_EXECUTABLE}" "print-cmakejs-src" "--log-level error" "--generator ${CMAKE_GENERATOR}"
+      WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+      OUTPUT_VARIABLE CMAKE_JS_SRC
+    )
+
+    execute_process(
+      COMMAND "${CMAKE_JS_EXECUTABLE}" "print-cmakejs-lib" "--log-level error" "--generator ${CMAKE_GENERATOR}"
+      WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+      OUTPUT_VARIABLE CMAKE_JS_LIB
+    )
+
+    # Strip the vars of any unusual chars that might break the paths...
+    _cmakejs_normalize_path(CMAKE_JS_INC)
+    _cmakejs_normalize_path(CMAKE_JS_SRC)
+    _cmakejs_normalize_path(CMAKE_JS_LIB)
+
+    string(REGEX REPLACE "[\r\n\"]" "" CMAKE_JS_INC "${CMAKE_JS_INC}")
+    string(REGEX REPLACE "[\r\n\"]" "" CMAKE_JS_SRC "${CMAKE_JS_SRC}")
+    string(REGEX REPLACE "[\r\n\"]" "" CMAKE_JS_LIB "${CMAKE_JS_LIB}")
+
+    set(CMAKE_JS_INC "${CMAKE_JS_INC}" CACHE STRING "cmake-js include directory." FORCE)
+    set(CMAKE_JS_SRC "${CMAKE_JS_SRC}" CACHE STRING "cmake-js source file." FORCE)
+    set(CMAKE_JS_LIB "${CMAKE_JS_LIB}" CACHE STRING "cmake-js lib file." FORCE)
+
+    # TODO: At this point, some warnings may occur re: the below (still investigating);
+    # Define either NAPI_CPP_EXCEPTIONS or NAPI_DISABLE_CPP_EXCEPTIONS.
+    #set (NAPI_CPP_EXCEPTIONS TRUE CACHE STRING "Define either NAPI_CPP_EXCEPTIONS or NAPI_DISABLE_CPP_EXCEPTIONS")
+    add_definitions(-DNAPI_CPP_EXCEPTIONS) # Also needs /EHsc
+    # add_definitions(-DNAPI_DISABLE_CPP_EXCEPTIONS)
```

Comment:
I don't think this belongs here, as different users will want different behaviours. I think we should leave this up to them to handle, as places such as the node-addon-api docs will tell them to set these as part of their setup guides and we don't want to be annoyingly opinionated